front/lib/plans/subscription.ts

```diff
@@ -0,0 +1,182 @@
+
+    if (activeSubscription.startDate >= today) {
+      await activeSubscription.update({
+        status: "cancelled",
+        endDate: today,
+      });
+    } else {
+      await activeSubscription.update({
+        status: "ended",
+        endDate: yesterday,
+      });
+    }
+  }
+
+  // We return the default subscription to FREE_TEST_PLAN
+  const freeTestPlan: PlanAttributes = FREE_TEST_PLAN_DATA;
+
+  return {
+    code: freeTestPlan.code,
+    name: freeTestPlan.name,
+    status: "active",
+    startDate: null,
+    endDate: null,
+    limits: {
+      assistant: {
+        isSlackBotAllowed: freeTestPlan.isSlackbotAllowed,
+        maxMessages: freeTestPlan.maxMessages,
+      },
+      connections: {
+        isSlackAllowed: freeTestPlan.isManagedSlackAllowed,
+        isNotionAllowed: freeTestPlan.isManagedNotionAllowed,
+        isGoogleDriveAllowed: freeTestPlan.isManagedGoogleDriveAllowed,
+        isGithubAllowed: freeTestPlan.isManagedGithubAllowed,
+      },
+      dataSources: {
+        count: freeTestPlan.maxNbStaticDataSources,
+        documents: {
+          count: freeTestPlan.maxNbStaticDocuments,
+          sizeMb: freeTestPlan.maxSizeStaticDataSources,
+        },
+      },
+      users: {
+        maxUsers: freeTestPlan.maxUsersInWorkspace,
+      },
+      largeModels: false, // TODO: remove this, it is always true now (kept to limit the scope of the PR)
+    },
+  };
+};
+
+/**
+ * Internal function to subscribe to the FREE_UPGRADED_PLAN.
+ * This plan is free with no limitations, and should be used for Dust workspaces only (once we have paiement plans)
+ */
+export const internalSubscribeWorkspaceToFreeUpgradedPlan = async ({
+  workspaceId,
+}: {
+  workspaceId: string;
+}): Promise<PlanType> => {
+  const workspace = await Workspace.findOne({
+    where: { sId: workspaceId },
+  });
+  if (!workspace) {
+    throw new Error(`Cannot find workspace ${workspaceId}`);
+  }
+  const plan = await Plan.findOne({
+    where: { code: FREE_UPGRADED_PLAN_CODE },
+  });
+  if (!plan) {
+    throw new Error(
+      `Cannot subscribe to plan ${FREE_UPGRADED_PLAN_CODE}:  not found.`
+    );
+  }
+
+  const today = new Date();
+
+  // We search for an active subscription for this workspace
+  const activeSubscription = await Subscription.findOne({
+    where: { workspaceId: workspace.id, status: "active" },
+  });
+  if (activeSubscription && activeSubscription.planId === plan.id) {
+    throw new Error(
+      `Cannot subscribe to plan ${FREE_UPGRADED_PLAN_CODE}:  already subscribed.`
+    );
+  }
+
+  return await front_sequelize.transaction(async (t) => {
+    // We end the active subscription if any
+    if (activeSubscription) {
+      const yesterday = new Date(today);
+      yesterday.setDate(yesterday.getDate() - 1);
+
+      if (activeSubscription.startDate >= today) {
+        await activeSubscription.update({
+          status: "cancelled",
+          endDate: today,
```

Comment:
Why not use `new Date()` here and in the startDate below? not sure I understand the value of having days vs times here?

Reply:
Discussed IRL -> I wanted to treat dates as days and have no overlap between subscription (hence the status cancelled if a subscription is created and ended the same day.

I removed all that and followed your reco.